### DISCUSSION TITLE
How to cancel a click/move event in a PlayerInventory

### DISCUSSION THREAD

**Discussion by bananasmoothii** [Minestom/Minestom#1317] [Help]
I've found that I have to use `InventoryPreClickEvent` to cancel an event in an `Inventory` (could you also explain me why `InventoryClickEvent` is not cancellable ?), but then I found that `PlayerInventory` does not extend `Inventory`, so using `InventoryPreClickEvent` doesn't work.

**Comment by GreatWyrm** [comment]
While PlayerInventory does not extend Inventory, PlayerInventory still calls associated InventoryClickProcessor methods that will call the PreClickEvents. You can still listen to PreClickEvents and cancel if the event.getInventory() is null (Player Inventory).

As for the exact design decision to have a PreClick and Click events, you will have to ask some of the devs about that, but think of it like this:
PreClickEvent - the client has sent a click to the server and wants to perform an action, you can cancel this to interrupt
ClickEvent - The click has been accepted by the server and actions will be performed

**Reply by bananasmoothii** [reply]
Hi, unfortunately `InventoryPreClickEvent` is not called at all when moving items in the inventory

**Reply by bananasmoothii** [reply]
It does get called when clicking on the player inventory while having another inventory open, otherwise it doesn't

**Reply by GreatWyrm** [reply]
It does get fired when moving items in the inventory, just checked with this piece of code:
```java
MinecraftServer.getGlobalEventHandler().addListener(InventoryPreClickEvent.class, event -> {
    System.out.println("Event fired, inventory: " + event.getInventory());
});
```
What does your PreClickEvent listener code look like?

**Reply by bananasmoothii** [reply]
I am registering my events through reflexion, and this doesn't seem to be the issue

https://user-images.githubusercontent.com/45853225/185661015-1274ef64-f349-4f81-9ceb-7c6997e35730.mp4

**Reply by GreatWyrm** [reply]
Ah, I see

IIRC, Creative mode is a bit weird with inventory stuff because of how it's implemented. Try seeing if the event fires in survival mode when clicking in your inventory.

**Reply by bananasmoothii** [reply]
indeed, it works in survival mode... I guess an issue can be created from this

**Reply by MrGazdag** [reply]
This is not a minestom bug, see my comment on #1343
